### ISSUE TITLE
Should PALs be entirely static &| can the inheritance diagram be simplified?

### ISSUE THREAD

**Issue by nwf** [microsoft/snmalloc#233]
Presently, some methods in PALs are not `static`: `notify_using`, `notify_not_using`, `zero`, `reserve_aligned`, `reserve_at_least`, `expensive_low_memory_check`, and possibly others I've missed.

That might be OK, but AIUI the PAL actually gets instantiated repeatedly and (IMHO) confusingly (I may be about to reveal that my understanding of C++ is somewhere between "stone age" and "stumbling"; if I'm overlooking something obvious, I apologize; I did at least attempt to verify my assertions with some strategically placed `std::cerr` debugging, but that may have kicked me into an entirely different basin of attraction within the C++ standard):

* Twice on the stack, inside `MemoryProviderStateMixin<PAL>::make()`: https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L85 

  * Once, since `MemoryProviderStateMixin<PAL>` extends `PAL`: https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L60

  * And *again*, since `MemoryProviderStateMixin<PAL>` has a member of type `AddressSpaceManger<PAL>` and that type extends `PAL`: https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L71 and https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/address_space.h#L17

* "The same" twice in that method again, this time in the nascent heap: https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L88-L89 more specifically at
https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L191

The on-heap `address_space` gets paved over with bits from the on-stack `address_space` using `::memcpy`: https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L102-L105
Ultimately, this means that only the on-heap `MemoryProviderStateMixin` parent-class `PAL` and the initially-on-stack `AddressSpace` parent class persist (well, the second of those as bits, anyway, not as a real object; the on-stack objects are now out of scope and destroyed).  The initially-on-hap `AddressSpace` object no longer exists in a sense, but it has had its bits paved over.

I don't know about you, but I find that at the very least more exciting than I might wish.  Making things even more exciting...

`MemoryProviderStateMixin` is inconsistent about which copy of the PAL it uses to achieve particular ends.  Anything that goes through the `AddressSpace<PAL> address_space` member will naturally use that PAL; examples include https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L188 which might chain through to the (potentially stateful) `reserve_aligned` at https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/address_space.h#L200  Internally, `MemoryProviderStateMixin` accesses its own parent class PAL at least once: https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L138  Meanwhile users of `MemoryProviderStateMixin` for PAL methods will hit its parent PAL as well, rather than that of the `address_space`, as happens in https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L237

FWIW, the OE PAL already takes its `heap_base` to be `static`, and the Windows PAL also considers its low-memory-notification state `static`.

So can we...

* insist that PALs be entirely `static` (updating those methods mentioned at the top), and/or
* make `AddressSpace` and/or `MemoryProviderStateMixin` not inherit from the PAL?

Or have I completely missed the mark?

**Comment by mjp41** [comment]
So one scenario we want is for there to be multiple Pals of the same type, but using different parts of the address range.  This is required for @davidchisnall's cross process allocation implementation for library compartmentalisation.  This means that there may be several Pal's for libraries in different parts of the address space, and thus each one requires its own state.  This unfortunately complicates stuff a lot.  And there is no testing for this scenario, which makes it hard to maintain.

Perhaps, you or David might be interested in adding a test for this?

**Comment by davidchisnall** [comment]
@nwf, [here](https://github.com/microsoft/verona/blob/a2f16c4500d819ddfb8aebbb2ce27aa49395c14e/experiments/process_sandbox/libsandbox.cc#L77) is an example of a stateful PAL.  It would be great if you could distil a test case for this.  #221 is pending merge waiting for a test for this use case.

**Comment by davidchisnall** [comment]
Actually, now that the `AddressSpaceManager` exists, that can probably be the stateful component.  It should be possible to refactor that experiment to use a shared `AddressSpaceManager` and leave the PAL stateless.

**Comment by nwf** [comment]
So, I've been doing some experimenting and have made a (very rough, tested only on one platform, &c) cut at decoupling the `PAL`, `AddressSpaceManager`, and (now somewhat erroneously named) `MemoryProviderStateMixin`; in this sketch, neither of the latter two inherit from the former, though there is a little delegation to keep the diff somewhat confined.  But I find myself somewhat lost in a maze of twisty passages.  I believe the following things of the desired design, any or all of which could be wrong:

* The `Pal` should be stateless, be globally available throughout `snmalloc`, and, on all platforms, offer `error`, `notify_using`, `notify_not_using`, `zero`, and `reserve_at_least` primitive operations (statically) as well as `static constexpr uint64_t page_size`.  The `notify*` primops may  be no-ops, but we expect the compiler to inline them away if so.

* Gated by `AlignedAllocation`, some `Pal`s also offer `reserve_aligned`, which is a perfectly fine static primop.

* The `AddressSpaceManager` is expected to be stateful, but its functionality is limited to splitting large blocks of address space down to smaller ones.  As now, it never expects allocations to come back to it: once allocated, forever allocated.  It is also agnostic to the commitment status of the blocks its managing, save that the first pointer-sized word of each but the last in any (always power of two) size class is committed to form the queue of blocks.

* In sandbox use cases, the `Pal`'s `reserve_at_least` and `reserve_aligned` operations likely return `nullptr` and the initial sandbox region is instead fed into the `AddressSpaceManager` as the sandbox is being initialized.  (Should we gate `reserve_at_least` behind another feature flag?)

* The `MemoryProviderStateMixin` (maybe just `MemoryProvider`) is also stateful, but, in cooperation with the various `LargeAlloc`s out in the world, manages only lockless queues of large-size-classed objects.  These objects never merge or split, but they may be decommitted on memory pressure by virtue of the `MemoryProviderStateMixin` acting as a `PalNotificationObject`.

Beyond those, I note that right now, the `Pal` is also responsible for memory pressure functionality, (i.e., `expensive_low_memory_check`, and `register_for_low_memory_callback`) but this is at odds with the desire that it be stateless.  On the other hand, this absolutely feels like a platform feature, or like something that will require platform abstraction.  I am tempted to move the (list of) `PalNotifier`(s) out to the `AddressSpaceManager` but leave the actual platform callback to the `Pal`.

**Comment by mjp41** [comment]
Two comments
* Being entirely `static` does not mean we are necessarily stateless?
* I am not convinced the Pal should be entirely stateless.  The Pal may track all kinds of state to simulate VM subsystem features.  Bits for pages that have been decommited, so we can lazily remove them.

I don't know if we really require the inheritance in the way we currently do: Having the first field of the `AddressSpaceManager` being a pointer to the `Pal`, and the first field of the `MemoryStateProvider` being a pointer to the `AddressSpaceManager` seems reasonable. Taking indirections on these paths will not cost anything noticable.

Having a consistent Pal we access is definitely something we should have.  It is not currently causing issues, but will with the future directions we are planning.

I think this issue is experiencing mission creep.  Let's focus on just making everything access a single stateful `Pal` initially. Remove the static accesses that aren't go through the appropriate one used by the `AddressSpaceManager`.

**Comment by davidchisnall** [comment]
I agree: we need to be careful about how we define state.  The PAL is intrinsically stateful, because it manages the state of the address mappings.  I believe the goal is for the PAL to have only state that relates to the global view of the address space.  This is either managed by the OS or maintained by the PAL.  In both cases, the PAL APIs should be entirely `static` and there should be no per-instance PAL state.  

The PAL maintains the platform state, the ASM maintains the state related to an allocator's view of the address space.  Does that make sense?

**Comment by mjp41** [comment]
Ultimately, a static Pal makes everything easier except extensions.  The complexity around stateful `Pal` and `MemoryProviderMixin` was added to support the compartmentalisation use case. If that really isn't require now, or in the future, then I am happy to go with making the Pal static.  However, I don't want us to remove it, to add back in a year.

The current `AddressSpaceManager` has a very well defined task: 

> produce nicely aligned memory from a given Pal. 

We could extend that to 

> produce nicely aligned memory from a given Pal or a fixed region.

and that would work fine.  So the fixed region would have a special pal that doesn't support `reserve_*`.

My concern about having the `Pal` being entirely static is if the compartmentalisation ever needs to do work specific to the compartment in the Pal.  For instance, if we need something to grow a compartment, then that is going to be stateful based on which compartment it is.  Your current design is a flat commonly mapped nGiB region, but if you have multiple regions for sharing with a process and these can be created by the Pal, then you need a per-compartment Pal.

**Comment by mjp41** [comment]
I think you have mostly resolved this issue with #245.  There is just the 

https://github.com/microsoft/snmalloc/blob/4e1f5829a754ab9c170ec090979d3a670d2d5d1a/src/mem/largealloc.h#L102-L105

That you mention that probably merits a separate issue, and we close this one off.
